File: clients/flutter-app/lib/labs/lab07_wysiwyg.dart
import 'dart:math' as math;

import 'package:colyseus/colyseus.dart';

import '../controls.dart';
import '../hud.dart';
import '../lab.dart';
import '../net/schema_bridge.dart';
import '../palette.dart';
import '../sim/sim.dart';
import 'move_lane.dart';
import '../gen/schema.dart';

/// Hitting something that is moving, and having the server agree.
///
/// The bot patrols, you drive into it, and the shove has to land the same
/// frame on both sides. That only works if both sides answer the same
/// question: *where was the bot at the instant this input was stamped?*
///
/// - The server rewinds the bot to that instant (`allowRewindState` with
///   `mode: "reckon"`, plus the per-input `lastSeenBy` read).
/// - The client asks [Predict.valueAt] for the same instant, [StepContext
///   .reckonTime]. Reading anything else — the newest snapshot, the smoothed
///   render position — tests against a pose the server never looks at.
/// - [StepContext.memoVec] freezes the verdict. The client keeps no history of
///   the bot, so a rollback replay would re-reckon it from a *newer* snapshot
///   and could flip a knife-edge call. The memo computes once, on the live
///   step, and replays that tuple verbatim.
///
/// What to look for: the marker where the hit was tested sits under the drawn
/// bot while the toggle is on. Turn it off and the marker jumps back onto the
/// dashed snapshot ring, roughly half a round trip behind — and grazes start
/// disagreeing with the server, which shows up as corrections that shove you
/// out of a bot you never touched.
///
/// `bumpTicks` is the other half: post-bump immunity is reconciled state, so
/// the countdown rewinds and replays with the position. "Can I be bumped right
/// now?" is therefore never itself a misprediction.
class Lab07Wysiwyg extends Lab {
  @override
  String get id => '07';

  @override
  String get title => 'WYSIWYG Collision';

  @override
  String get blurb => 'Client and server test the hit at the same instant.';

  /// Verdicts kept for the replay check. Comfortably more than the in-flight
  /// window, which is what bounds how many inputs can be replayed.
  static const _verdictLogMax = 128;

  /// A correction this big right after a predicted bump is a flipped verdict,
  /// not knockback-direction epsilon: the 48 u/s shove either happened or it
  /// did not.
  static const _flippedCorrection = 3.0;

  MoveLane<BumpPlayer>? lane;
  final StepBinding _botBinding = StepBinding();

  /// Read the bot at `ctx.reckonTime` (the server's instant) or at the newest
  /// snapshot (the naive read). Live — the step reads it per call.
  bool _atReckonTime = true;

  /// Verdicts frozen on the live step, keyed by input sequence, so a replay of
  /// the same input can be compared against what the live step decided.
  final Map<int, List<double>> _verdicts = {};

  /// Bumps this client predicted.
  int bumps = 0;

  /// Replayed inputs whose verdict was checked against the live one.
  int replayedVerdicts = 0;

  /// Replays that reached a different verdict. Stays at 0 while the memo is
  /// doing its job.
  int verdictFlips = 0;

  /// Corrections large enough, and soon enough after a bump, to be a verdict
  /// the server disagreed with.
  int mispredicts = 0;

  int _lastReconcileSeq = 0;
  double _lastBumpAt = double.negativeInfinity;
  double _flashAt = double.negativeInfinity;

  /// Server instant the newest live verdict was taken at.
  double _reckonTime = 0;

  /// Where that verdict was evaluated — the marker the lab is built around.
  double _evalX = double.nan;
  double _evalY = double.nan;

  /// The decoded `kind`, handed to the reckon step: its scratch carries
  /// scalars only, so a string field does not survive into it.
  String _kind = '';

  @override
  Future<bool> mount(LabContext ctx) async {
    // bumpTicks is reconciled state: the immunity counter has to rewind with
    // the position, or a rollback re-opens the gate and bumps twice.
    final joined = await MoveLane.mount(
      ctx.client,
      playerType: BumpPlayer.new,
      roomName: 'lab-bump',
      fields: const ['x', 'y', 'vx', 'vy', 'bumpTicks'],
    );
    if (joined == null) return false;

    lane = joined;
    room = joined.room;

    // Reckoned rather than interpolated: valueAt can only answer for an
    // arbitrary instant when the field is projected through the shared step.
    joined.predict
        .attachAllReckon('bots', const ['x', 'y'], _reckonBot, smoothMs: 40);
    joined.onStep = _step;
    return true;
  }

  /// The forward pass the SDK runs over a scratch copy of the bot's snapshot.
  void _reckonBot(SchemaView state, double dt, double elapsedMs) {
    stepBot(_botBinding.body(state), dt, elapsedMs, kindOverride: _kind);
  }

  /// The bots collection, re-read every time — the decoder replaces instances.
  MapSchema<Bot>? get bots => room?.stateAs(BumpState.new)?.bots;

  /// The patrolling bot, for readouts.
  Bot? get bot => bots?['bot1'];

  void _step(StepContext ctx, SchemaBody body) {
    // Only a countdown, so running it after the movement step lands on the
    // same number the server gets running it before — and it still precedes
    // the collision test, which is what reads it.
    stepBumpGate(body);

    final when = ctx.reckonTime;
    final verdict = ctx.memoVec('bump', () => _testBots(body, when));
    _trackReplay(ctx, verdict);
    if (verdict.length < 3 || verdict[0] == 0) return;

    body.vx = verdict[1];
    body.vy = verdict[2];
    body.bumpTicks = bumpCooldownTicks; // immunity rides adopt + replay
    if (ctx.isReplay) return; // counters and the flash: live step only

    bumps++;
    _lastBumpAt = lane?.room.clock.now ?? 0;
    _flashAt = _lastBumpAt;
  }

  /// The bump verdict as a memo-able tuple: `[hit 0/1, vx, vy]`.
  ///
  /// Called from inside the memo, so it runs once per input on the live step —
  /// which is also why recording the evaluated position here is safe: a replay
  /// never re-enters it.
  List<double> _testBots(SchemaBody player, double when) {
    final predict = lane?.predict;
    final all = bots;
    if (predict == null || all == null) return const [0, 0, 0];

    _reckonTime = when;
    for (final entry in all.entries) {
      final target = entry.value;

      // The server rewinds THIS input to `when`, so this is the only read that
      // can agree with it. The naive branch takes the newest decoded sample
      // instead: about half a round trip behind what the server tests.
      final bx =
          _atReckonTime ? predict.valueAt(target, 'x', when) : target.x;
      final by =
          _atReckonTime ? predict.valueAt(target, 'y', when) : target.y;
      _evalX = bx;
      _evalY = by;

      final hit = collideBot(player, bx, by);
      if (hit != null) return [1, hit.vx, hit.vy];
    }
    return const [0, 0, 0];
  }

  /// Compares a replayed input's verdict against the one the live step froze.
  ///
  /// This is the memo's whole value made visible: without it a replay would
  /// re-derive the test against fresher bot data, and [verdictFlips] would
  /// start climbing.
  void _trackReplay(StepContext ctx, List<double> verdict) {
    if (!ctx.isReplay) {
      if (_verdicts.length >= _verdictLogMax) {
        _verdicts.remove(_verdicts.keys.first);
      }
      _verdicts[ctx.tick] = verdict;
      return;
    }

    final live = _verdicts[ctx.tick];
    if (live == null) return;
    replayedVerdicts++;
    if (live.length != verdict.length) {
      verdictFlips++;
      return;
    }
    for (var i = 0; i < live.length; i++) {
      if (live[i] == verdict[i]) continue;
      verdictFlips++;
      return;
    }
  }

  @override
  void frame(LabContext ctx, double now, double dtMs) {
    final l = lane;
    if (l == null || !l.ready) return;

    final target = bot;
    if (target != null) _kind = (target['kind'] as String?) ?? '';

    l.drive(now);

    final recon = l.reconciler;
    if (recon == null || recon.reconcileSeq == _lastReconcileSeq) return;
    _lastReconcileSeq = recon.reconcileSeq;
    if (recon.lastCorrectionMag <= _flippedCorrection ||
        now - _lastBumpAt >= 700) {
      return;
    }
    mispredicts++;
    // One per bump: a flipped verdict keeps correcting for several acks while
    // the in-flight window drains, and that is still one disagreement.
    _lastBumpAt = double.negativeInfinity;
  }

  @override
  void render(LabContext ctx) {
    final l = lane;
    if (l == null) return;

    final draw = ctx.draw;
    final view = ctx.view;

    for (final other in l.others) {
      draw.square(
          other.x, other.y, playerHalf, hueColor(other.hue, alpha: 0.45));
    }

    final all = bots;
    if (all != null) {
      for (final entry in all.entries) {
        final target = entry.value;

        // Where the bot is drawn: forward-simulated to the present.
        final bx = l.predict.value(target, 'x');
        final by = l.predict.value(target, 'y');
        draw.circle(bx, by, botRadius, Palette.warn.fade(0.28));
        draw.circleOutline(bx, by, botRadius, Palette.warn, width: 2);
        draw.label(bx, by, 'bot (drawn)', Palette.warn,
            size: 10, dy: -view.s(botRadius) - 14);

        // The newest decoded sample, which is already stale by the time it
        // arrives. The gap to the ring above is what reckoning buys back.
        final rawX = target.x;
        final rawY = target.y;
        draw.circleOutline(rawX, rawY, botRadius, Palette.text.fade(0.4),
            width: 1, dashed: true);
        draw.label(rawX, rawY, 'latest snapshot', Palette.text.fade(0.4),
            size: 9, dy: view.s(botRadius) + 4);
      }
    }

    // Where the collision was actually evaluated. On reckonTime it lands under
    // the drawn bot; on the snapshot it lands under the dashed ring, and that
    // is a position the server never tests against.
    if (!_evalX.isNaN) {
      final tone = _atReckonTime ? Palette.accent : Palette.bad;
      draw.marker(_evalX, _evalY, botRadius * 0.6, tone);
      draw.label(_evalX, _evalY, 'hit tested here', tone,
          size: 9, dy: view.s(botRadius) + 16);
    }

    // Where the server last said we were, against where we predict we are.
    draw.ghostSquare(l.serverX, l.serverY, playerHalf, Palette.text.fade(0.5));

    final flash = (l.room.clock.now - _flashAt) < 300;
    draw.square(l.predictedX, l.predictedY, playerHalf,
        flash ? Palette.bad : hueColor(l.hue));
    draw.squareOutline(l.predictedX, l.predictedY, playerHalf,
        Palette.text.fade(flash ? 1 : 0.85),
        width: flash ? 2.5 : 1.5);

    _renderHud(ctx, l);
  }

  void _renderHud(LabContext ctx, MoveLane<BumpPlayer> l) {
    final hud = ctx.hud;
    final clock = l.room.clock;
    final recon = l.reconciler;

    hud.section('COLLISION');
    hud.row('bumps predicted', '$bumps');
    hud.row('bumps (server)', '${(l.me?['bumps'] as num?)?.toInt() ?? 0}');
    hud.row('immunity (ticks)',
        '${(recon?.state['bumpTicks'] ?? 0).round()}');
    final rate = bumps == 0 ? 0.0 : mispredicts / bumps * 100;
    hud.row(
      'mispredicts',
      bumps == 0 ? '$mispredicts' : '$mispredicts (${rate.round()} %)',
      tone: mispredicts == 0 ? HudTone.good : HudTone.warn,
    );

    hud.section('RECKONING');
    // The instant the verdict was taken at, relative to now and to the newest
    // snapshot: the first is how far back the test looked, the second is how
    // far the bot had to be projected forward to answer.
    final offset = math.max(0.0, clock.serverNow - _reckonTime);
    final horizon = _reckonTime - clock.lastServerTime;
    hud.row('reckon offset (now - eval)', '${offset.round()} ms');
    hud.row('projected past snapshot', '${horizon.round()} ms');
    hud.row('eval vs snapshot', '${_evalGap().toStringAsFixed(1)} u');
    hud.row('read', _atReckonTime ? 'ctx.reckonTime' : 'latest snapshot',
        tone: _atReckonTime ? HudTone.good : HudTone.bad);

    hud.section('REPLAY');
    hud.row('verdicts replayed', '$replayedVerdicts');
    hud.row('verdicts that changed', '$verdictFlips',
        tone: verdictFlips == 0 ? HudTone.good : HudTone.bad);

    hud.section('PREDICTION');
    hud.chips(l.pending, label: 'pending');
    final status = classifyDrift(l.drift);
    hud.row('drift', status.name, tone: switch (status) {
      DriftStatus.matched => HudTone.good,
      DriftStatus.jitter => HudTone.warn,
      DriftStatus.diverging => HudTone.bad,
    });
    hud.row('drift ema', l.drift.ema.toStringAsFixed(4));
    hud.row('rtt', '${clock.smoothedRtt.round()} ms');

    hud.note(_atReckonTime
        ? 'The verdict is taken at the instant the server rewinds this input '
            'to, and frozen with memoVec so a rollback replays it instead of '
            're-deriving it. Both counters above stay flat because the two '
            'sides are answering the same question.'
        : 'The verdict now comes from the newest snapshot, which is about half '
            'a round trip behind the pose the server tests. Graze the patrol '
            'path at high latency: bumps the server never awarded, and misses '
            'it did, both arrive as corrections.');
  }

  /// Distance between the evaluated pose and the newest snapshot, in world
  /// units — the visible size of the disagreement the naive read introduces.
  double _evalGap() {
    final target = bot;
    if (target == null || _evalX.isNaN) return 0;
    final dx = _evalX - target.x;
    final dy = _evalY - target.y;
    return math.sqrt(dx * dx + dy * dy);
  }

  @override
  List<ControlSpec> controls(LabContext ctx) {
    return [
      ToggleSpec('evaluate at ctx.reckonTime', _atReckonTime, (v) {
        _atReckonTime = v;
        mispredicts = 0;
        bumps = 0;
        verdictFlips = 0;
        replayedVerdicts = 0;
      }),
      const NoteSpec('On, the bot is read at the instant the server rewinds '
          'this input to, so both sides test the same pose. Off, the newest '
          'snapshot is read instead. Raise the latency preset and graze the '
          'patrol path: head-on hits survive a stale read, grazes do not.'),
    ];
  }

  @override
  void onReconnect() {
    // Sequences restart at zero, so the frozen verdicts belong to inputs that
    // no longer exist.
    _verdicts.clear();
    _lastReconcileSeq = 0;
    replayedVerdicts = 0;
    verdictFlips = 0;
  }

  @override
  void unmount() {
    lane?.dispose();
    lane = null;
    _verdicts.clear();
  }
}

File: clients/native-app/labs/lab07_wysiwyg.c
/*
 * Lab 07 — WYSIWYG Collision.
 *
 * The collision against the moving bot is predicted INSIDE the reconciler
 * step. Two ingredients make the client's verdict equal the server's:
 *
 *  1. predict_value_at(bot, "x", ctx->reckon_time) — read the bot at the
 *     instant the server will REWIND this input to. Reading anything else
 *     (the stale snapshot, or the smoothed render pose) tests against a
 *     position the server never uses.
 *
 *  2. colyseus_step_memo(...) — the verdict is NOT re-derivable on rollback
 *     replay (the client keeps no bot history; a later replay would read the
 *     bot reckoned from a NEWER snapshot and could flip the call). memo runs
 *     the test once on the live step and replays the outcome verbatim.
 *
 * The verdict is a VECTOR (the knockback vx/vy), so it rides one
 * colyseus_step_memo_vec: the collision test runs exactly once on the live step
 * and both components replay together. Encoding it as an angle would round
 * through atan2/cos and reintroduce the drift this lab exists to eliminate.
 *
 * Port of src/client/labs/07-wysiwyg/.
 */

static struct {
    colyseus_room_t* room;
    bump_state_t* state;
    const char* sid;

    colyseus_predict_t* predict;
    colyseus_input_handle_t* input;
    move_input_t* cmd;

    bump_player_t* me;
    colyseus_reconciler_t* recon;
    bump_player_t* predicted;
    bool rebind;

    bot_t* bot;
    bool use_value_at;
    bool use_memo;

    int bumps_predicted;
    double last_bump_at, bump_flash_t;
    int mispredicts;
    int last_reconcile_seq;
} l07;

/* --------------------------------------------------- the bump verdict */

typedef struct {
    double px, py;
    int bump_ticks;
    double when;
} l07_probe_t;

typedef struct { const l07_probe_t* probe; bool hit; double vx, vy; } l07_scan_t;

static void l07_scan_bot(const char* key, void* value, void* userdata) {
    (void)key;
    l07_scan_t* s = (l07_scan_t*)userdata;
    if (s->hit) { return; }
    colyseus_schema_t* bot = (colyseus_schema_t*)value;
    double bx, by;
    if (l07.use_value_at) {
        /* The server rewinds THIS input to reckon_time — test there. */
        bx = colyseus_predict_value_at(l07.predict, bot, "x", s->probe->when);
        by = colyseus_predict_value_at(l07.predict, bot, "y", s->probe->when);
    } else {
        bx = ((bot_t*)bot)->x;   /* the wrong read: ~RTT/2 behind the server */
        by = ((bot_t*)bot)->y;
    }
    s->hit = collide_bot(s->probe->px, s->probe->py, s->probe->bump_ticks,
        bx, by, &s->vx, &s->vy);
}

/* Writes the knockback into out[0..1]; returns 0 for "no bump this step". */
static int l07_test_bots(double* out, void* userdata) {
    const l07_probe_t* probe = (const l07_probe_t*)userdata;
    l07_scan_t scan = { probe, false, 0, 0 };
    colyseus_map_schema_foreach(l07.state->bots, l07_scan_bot, &scan);
    if (!scan.hit) { return 0; }
    out[0] = scan.vx;
    out[1] = scan.vy;
    return 2;
}

static void l07_step(const colyseus_step_ctx_t* ctx, colyseus_schema_t* state,
    const colyseus_schema_t* command, void* userdata) {
    (void)userdata;
    bump_player_t* p = (bump_player_t*)state;
    const move_input_t* inp = (const move_input_t*)command;

    int ticks = p->bumpTicks;
    step_bump_gate(&ticks);            /* reconciled tick gate (both sides) */
    p->bumpTicks = (uint8_t)ticks;

    entity_state_t e = { p->x, p->y, p->vx, p->vy };
    step_entity(&e, (double)inp->moveX, (double)inp->moveY, ctx->dt);
    p->x = e.x; p->y = e.y; p->vx = e.vx; p->vy = e.vy;

    l07_probe_t probe = { p->x, p->y, p->bumpTicks, ctx->reckon_time };
    double knock[COLYSEUS_MEMO_VEC_MAX];
    int hit = l07.use_memo
        ? colyseus_step_memo_vec(ctx, "bump", l07_test_bots, &probe, knock)
        : l07_test_bots(knock, &probe);

    if (hit == 2) {
        p->vx = knock[0];
        p->vy = knock[1];
        p->bumpTicks = BUMP_COOLDOWN_TICKS;   /* immunity rides adopt+replay */
        if (!ctx->is_replay) {                /* FX/counters: live step only */
            l07.bumps_predicted++;
            l07.last_bump_at = nd_now();
            l07.bump_flash_t = l07.last_bump_at;
        }
    }
}

/* ------------------------------------------------------------- setup */

static void l07_reckon_step(colyseus_schema_t* state, double dt, double elapsed_ms, void* userdata) {
    (void)userdata;
    bot_t* b = (bot_t*)state;
    bot_sim_t s = {
        b->x, b->y, b->vx, b->vy, b->kind,
        b->minX, b->maxX, b->baseY, b->phaseMs, b->speed, b->lastTeleport,
    };
    step_bot(&s, dt, elapsed_ms);
    b->x = s.x; b->y = s.y; b->vx = s.vx; b->vy = s.vy;
    b->lastTeleport = s.last_teleport;
}

static bool l07_make_reconciler(void) {
    static const char* const FIELDS[] = { "x", "y", "vx", "vy", "bumpTicks" };
    colyseus_reconciler_options_t opts = { 0 };
    opts.smooth_ms = 65;
    opts.fields = FIELDS;
    opts.field_count = 5;
    l07.recon = colyseus_predict_reconciler(l07.predict, (colyseus_schema_t*)l07.me,
        &bump_player_vtable, l07.input, l07_step, &opts);
    if (!l07.recon) { return false; }
    l07.predicted = (bump_player_t*)colyseus_reconciler_state(l07.recon);
    l07.last_reconcile_seq = 0;
    return true;
}

static bool lab07_attach(app_t* app, colyseus_room_t* room) {
    bump_state_t* state = (bump_state_t*)colyseus_room_get_state(room);
    if (!state || !state->players || !state->bots) { return false; }
    const char* sid = colyseus_room_get_session_id(room);
    bump_player_t* me = (bump_player_t*)colyseus_map_schema_get(state->players, sid);
    bot_t* bot = (bot_t*)colyseus_map_schema_get(state->bots, "bot1");
    if (!me || !bot) { return false; }

    memset(&l07, 0, sizeof(l07));
    l07.room = room;
    l07.state = state;
    l07.sid = sid;
    l07.me = me;
    l07.bot = bot;
    l07.use_value_at = true;
    l07.use_memo = true;
    l07.last_bump_at = -1e9;
    l07.bump_flash_t = -1e9;

    l07.predict = colyseus_predict_for_room(room);
    /* Bots are DEAD-RECKONED through the shared step — the timeline the
     * collision test below reads at ctx->reckon_time. */
    colyseus_predict_attach_all_reckon(l07.predict, (colyseus_schema_t*)state, "bots",
        &bot_vtable, SMOOTHED_XY, 2, l07_reckon_step, 25, 0, 0, NULL);
    colyseus_predict_attach_all(l07.predict, (colyseus_schema_t*)state, "players",
        ATTACH_XY_DAMPED, 2, sid, NULL);

    l07.input = colyseus_room_input(room, &move_input_vtable, NULL);
    if (!l07.input) { return false; }
    l07.cmd = (move_input_t*)colyseus_input_handle_data(l07.input);
    (void)app;
    if (!l07_make_reconciler()) { return false; }
    app_set_latency_preset(2);   /* a knife-edge verdict needs a real RTT */
    return true;
}

static void l07_draw_remote(const char* sid, void* value, void* userdata) {
    app_t* app = (app_t*)userdata;
    if (strcmp(sid, l07.sid) == 0) { return; }
    bump_player_t* p = (bump_player_t*)value;
    draw_square(&app->view,
        colyseus_predict_value(l07.predict, (colyseus_schema_t*)p, "x"),
        colyseus_predict_value(l07.predict, (colyseus_schema_t*)p, "y"),
        PLAYER_HALF, hue_color(p->hue, 0.4));
}

static void lab07_frame(app_t* app, double now, double dt) {
    (void)dt;
    if (l07.rebind) {
        bump_player_t* me = (bump_player_t*)colyseus_map_schema_get(l07.state->players, l07.sid);
        if (me) {
            if (l07.recon) { colyseus_reconciler_free(l07.recon); l07.recon = NULL; }
            l07.me = me;
            l07.bot = (bot_t*)colyseus_map_schema_get(l07.state->bots, "bot1");
            l07.rebind = !l07_make_reconciler();
        }
        if (!l07.recon) { return; }
    }

    /* Both toggles rebuild nothing — the step reads them live. */
    if (app_key(KEY_V)) { l07.use_value_at = !l07.use_value_at; }
    if (app_key(KEY_M)) { l07.use_memo = !l07.use_memo; }


    /* The acceptance script can't hand-steer into a moving bot; when it drives,
     * seek the bot's lane and let the patrol sweep do the rest. */
    int move_x = kb_move_x(), move_y = kb_move_y();
    if (g_autopilot) {
        double lane = colyseus_predict_value(l07.predict, (colyseus_schema_t*)l07.bot, "y");
        double dy = lane - l07.predicted->y;
        move_x = 0;
        move_y = dy > 1.0 ? 1 : dy < -1.0 ? -1 : 0;
    }

    int steps = colyseus_predict_tick(l07.predict, now);
    for (int i = 0; i < steps; i++) {
        l07.cmd->moveX = (int8_t)move_x;
        l07.cmd->moveY = (int8_t)move_y;
        colyseus_input_handle_send(l07.input);
    }

    /*
     * Attribute LARGE corrections near a predicted bump to a flipped verdict.
     * Small post-bump corrections (< ~3u) are knockback-direction epsilon; a
     * genuinely flipped verdict diverges by many units — the 48 u/s shove
     * either happened or it didn't.
     */
    int seq = colyseus_reconciler_reconcile_seq(l07.recon);
    if (seq != l07.last_reconcile_seq) {
        l07.last_reconcile_seq = seq;
        if (colyseus_reconciler_last_correction_mag(l07.recon) > 3
            && nd_now() - l07.last_bump_at < 700) { l07.mispredicts++; }
    }

    /* --- render ------------------------------------------------------ */
    const world_view_t* v = &app->view;
    colyseus_map_schema_foreach(l07.state->players, l07_draw_remote, app);

    double bx = colyseus_predict_value(l07.predict, (colyseus_schema_t*)l07.bot, "x");
    double by = colyseus_predict_value(l07.predict, (colyseus_schema_t*)l07.bot, "y");
    draw_circle_world(v, bx, by, BOT_RADIUS, with_alpha(COL_WARN, 0.3));
    draw_circle_outline_world(v, bx, by, BOT_RADIUS, COL_WARN);
    draw_label_world(v, bx, by, "bot (reckon = hit position)", COL_WARN, 10,
        -vs(v, BOT_RADIUS) - 14);
    draw_circle_dashed_world(v, l07.bot->x, l07.bot->y, BOT_RADIUS, with_alpha(COL_TEXT, 0.45));
    draw_label_world(v, l07.bot->x, l07.bot->y, "stale snapshot", with_alpha(COL_TEXT, 0.45), 9,
        vs(v, BOT_RADIUS) + 4);

    draw_ghost_square(v, l07.me->x, l07.me->y, PLAYER_HALF, with_alpha(COL_TEXT, 0.5));
    bool flash = nd_now() - l07.bump_flash_t < 300;
    double px = colyseus_reconciler_value(l07.recon, "x");
    double py = colyseus_reconciler_value(l07.recon, "y");
    draw_square(v, px, py, PLAYER_HALF, flash ? COL_BAD : hue_color(l07.me->hue, 1));
    draw_square_outline(v, px, py, PLAYER_HALF, COL_TEXT, flash ? 2.5f : 1.0f);

    /* --- HUD --------------------------------------------------------- */
    hud_t* h = &app->hud;
    hud_section(h, "TELEMETRY");
    hud_row(h, "bumps predicted", COL_TEXT, "%d", l07.bumps_predicted);
    hud_row(h, "bumps (authoritative)", COL_TEXT, "%d", l07.me->bumps);
    hud_row(h, "mispredicts", l07.mispredicts ? COL_WARN : COL_GOOD, "%d", l07.mispredicts);
    double rate = l07.bumps_predicted ? (double)l07.mispredicts / l07.bumps_predicted * 100 : 0;
    if (l07.bumps_predicted) {
        hud_row(h, "mispredict rate", rate < 10 ? COL_GOOD : rate < 40 ? COL_WARN : COL_BAD,
            "%.0f %%", rate);
    } else {
        hud_row(h, "mispredict rate", COL_TEXT_FAINT, "--");
    }
    hud_row(h, "bump immunity (ticks)", COL_TEXT, "%d", l07.predicted->bumpTicks);

    hud_section(h, "CONTROLS");
    hud_key(h, "WASD", "graze the bot's patrol path");
    hud_key(h, "V", l07.use_value_at ? "read bot at ctx.reckonTime" : "read the STALE snapshot");
    hud_key(h, "M", l07.use_memo ? "verdict frozen with memo" : "verdict RE-DERIVED on replay");
    hud_note(h, "Raise the latency preset and graze the bot repeatedly. V off tests against "
        "a position ~RTT/2 behind where the server tests; M off lets replays re-derive the "
        "verdict against newer bot data and flip knife-edge calls. Both off is maximally wrong.");
}

static void lab07_detach(app_t* app) {
    (void)app;
    if (l07.recon) { colyseus_reconciler_free(l07.recon); }
    if (l07.predict) { colyseus_predict_free(l07.predict); }
    memset(&l07, 0, sizeof(l07));
}

static void lab07_on_reconnect(app_t* app) { (void)app; l07.rebind = true; }

static const lab_def_t LAB_07 = {
    .id = "07-wysiwyg",
    .num = 7,
    .title = "WYSIWYG Collision",
    .blurb = "valueAt(reckonTime) + ctx.memo.",
    .room_name = "lab-bump",
    .state_vtable = &bump_state_vtable,
    .attach = lab07_attach,
    .frame = lab07_frame,
    .detach = lab07_detach,
    .on_reconnect = lab07_on_reconnect,
};
